airflow/providers/amazon/aws/operators/s3.py

```diff
@@ -318,6 +318,94 @@ def execute(self, context: 'Context'):
         )
 
 
+class S3CreateObjectOperator(BaseOperator):
+    """
+    Creates a new object from a given string or bytes.
```

Comment:
Given from where?
Can you describe a use case where this operator is needed?

Reply:
I updated the comment.

Well the common use case would be I want to upload a file to S3 (JSON, CSV, ...) and specify its content. Example:
- I pull some data from an API (e.g.) and I want to store the result into S3

Reply:
There is one DAG example in this PR. Please take a look at `example_s3.py`. I personally do not really like `LocalFilesystemToS3Operator` because it requires having a local file. If we take the same example I stated, it means I would need to create a local file after I fetched the data from the API and before I store it in S3. I would also need to remove that file afterwards

Reply:
No, the example dag in the PR is just how to set the Operator.
I want to see how you thought the operator is to be used with an actual workflow.

I think there may be a wrong assumption here.
Tasks do not share data between them (other than metadata via Xcom).

What I fail to understand is how this operator will get the given string? (Assuming the string is not hard coded)?
Would the string be retrieved from other operator? How?

Reply:
That's a good point. What I can propose is to add a parameter to retrieve data from Xcom. Here is an example of operator using that technique: https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/transfers/google_api_to_s3.py#L57

What do you think?

Reply:
I don't think it's the same...
It allows to retrieve configurations for the API call from Xcom. The is considered as metadata.
The operator you shared is `GoogleApiToS3Operator`
it reads from API and write to S3.
The xcom part you were referencing is prior to the API call... it's just about the configs

Reply:
What I meant was just to expose a parameter to get data from Xcom and upload that data into S3

Reply:
But for that to work it means that by some other way someone stored data in Xcom and then you want to create S3 object from the data stored in xcom.

This is why I asked to see an actual DAG example using this operator as part of a complete workflow.

Lets see what other thinks...

Reply:
> What I fail to understand is how this operator will get the given string? (Assuming the string is not hard coded)?

From within Airflow use cases I know of (see [here](https://github.com/apache/airflow/blob/2400de2c5ece644cadb870baeea28907fa4dcf58/airflow/providers/amazon/aws/example_dags/example_s3_to_redshift.py#L36), [here](https://github.com/apache/airflow/blob/2400de2c5ece644cadb870baeea28907fa4dcf58/airflow/providers/amazon/aws/example_dags/example_athena.py#L44) and [here](https://github.com/apache/airflow/blob/2400de2c5ece644cadb870baeea28907fa4dcf58/airflow/providers/amazon/aws/example_dags/example_glue.py#L75)) are from hard coded string/data in the dag file.

From user dags, I've seen both hardcoded and runtime.

> But for that to work it means that by some other way someone stored data in Xcom and then you want to create S3 object from the data stored in xcom. This may encourage bad practices.

What is bad practice about a workflow which consumes output of one operation and then writes that data to S3? Taking some output (whether that be json, text, csv, etc) from one operation and persisting that data to object storage is a very common pipeline workflow. In the S3 case you either must write an unnecessary/temporary file to disk or use the S3Hook directly. I've used both, and the latter is much cleaner, but not as convenient as a dedicated operator would be.

But that's just my 2c

Reply:
FWIW, I just ran into a similar situation working with the SageMaker operators where this would be handy.  Use case:  Given a URL, fetch the csv located there and upload it to S3.  I can either use the S3Hook().load_string() as in Niko's examples, or maybe write the data to a tempfile and use the LocalToS3 transfer; both should work fine, but the operator would be convenient.

```
@task
def load_data_to_S3():
     dataset = requests.get(DATA_URL).content
     s3_hook = S3Hook()
     s3_hook.load_string(dataset, f'{S3_KEY}/input_data.csv', S3_BUCKET, replace=True)
```

could become

```
 load_data_to_S3 = S3CreateObjectOperator(
     bucket=S3_BUCKET,
     key=f'{S3_KEY}/input_data.csv',
     data=requests.get(DATA_URL).content,
     replace=True,
 )
 ```
 
 I guess in the end maybe neither is strictly "better" but I'd say it isn't unreasonable for a user to expect to find an operator to do that.

Reply:
@ferruzzi The difference is that in your PythonOperator example `requests.get(DATA_URL)` is inside the python callable thus executed only when the task is executed while in the `S3CreateObjectOperator` the `requests.get(DATA_URL)` will be executed every time the dag is parsed.

What you wrote should be in fact`HttpToS3Operator` which is a transfer operator. The `S3CreateObjectOperator` as shown here is not transfer operator.

Reply:
`HttpToS3Operator` doesn't exist (yet) but I can see how that would be a solution.  Your correction about the timing of the execution is right though, and appreciated.  Sorry about that.

Reply:
So @ferruzzi can you describe/show a use case of using the operator as implemented in this PR?

Reply:
I guess from this discussion we can close this PR and instead, create another to create `HttpToS3Operator `?

Reply:
I'm not sure if you saw my reply @eladkal, but in those three cases I linked, such an http to S3 operator would not help. 

Another example, that is not of the hardcoded variety, that I've run into recently was with the Step Functions operator. There is an operator to run a SF workflow and then another to fetch the json result which is returned, it would be convenient to have an S3 operator to consume the output from xcom and write it to S3, instead of having to use the python operator and write boilder plate code that uses the hook directly to create the object or create a tempfile and use the local to s3 operator.

I'm not sure what the grave concern is about adding this operator, but I'd sure like to have it, and would make use of it.

Maybe we can flip the question around and here what concerns, risks, oversights and pitfalls you see over including this operator? Thanks!

Reply:
> Maybe we can flip the question around and here what concerns, risks, oversights and pitfalls you see over including this operator? Thanks!

None.

I said that my concern is the use case of how this operator is to be used. I'm concerned that users will use this differently than what intended - and we actually discovered during this discussion that the original intention was for 
`HttpToS3Operator`. The process we did here is not something users are likely to do. I am more worried that we are serving users a way to a bad practice even if we don't mean that.

I'd be more comfortable with adding example dag which emphasise how the operator is expected to be used (e.g an actual example that shows the data expected to be passed is metadata).

That said - I am not going to block this operator. We were just having discussion.

If you still think differently i'd be happy to resolve this discussion and more to the code review part.

Reply:
>  and we actually discovered during this discussion that the original intention was for HttpToS3Operator


I think the conversation got a bit muxed by one example being given at the top then an assumption being made that such an example is the only applicable usecase ever possible. I have mentioned examples in my previous two comments about usecases for which the HttpToS3Operator would _not_ help and this operator would. But, as always, I'm happy to disagree a commit if you deem it dangerous or reckless to merge this than we can continue to use the hook directly.

Reply:
If you can bake one of these examples into the example dag of this operator it's enough for me :)

Reply:
I'm resolving this thread and let's move to code review.
Should @vincbeck be able to create such example dag that would be great.

Reply:
+1 to showing usage in the example dag for users.

Another way to show its usage would be updating some of the code I linked which uses the load_string hook method directly.